repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -51,7 +54,9 @@ class Session(interpreter: Interpreter)
   private implicit val formats = DefaultFormats
 
   private var _state: SessionState = SessionState.NotStarted()
-  private var _history = IndexedSeq[Statement]()
+  private val _statements = mutable.Map[Int, Statement]()
```

Comment:
I would suggest to use guava's cache instead of this `Map`, that will be easy for you to later support retention.